### PR TITLE
Add MS SQL Server APIs 

### PR DESCRIPTION
**Pull request by Neaj-Morshad-101**

### CODE REVIEW

File: apis/catalog/v1alpha1/mssql_version_helpers.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+func (m MsSQLVersion) ResourceSingular() string {
+	return ResourceSingularMsSQLVersion
+}
+
+func (m MsSQLVersion) ResourcePlural() string {
+	return ResourcePluralMsSQLVersion
+}
+
+func (m MsSQLVersion) ValidateSpecs() error {
+	if m.Spec.Version == "" || m.Spec.DB.Image == "" || m.Spec.Coordinator.Image == "" {
```
Comment (souravbiswassanto) [inline apis/catalog/v1alpha1/mssql_version_helpers.go:56]: TODO for you, add m.spec.exporter.image check too when you work with monitoring

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+				".spec.topology.mode can't be empty in cluster mode"))
+		}
+
+		if m.Spec.Replicas == nil {
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+				m.Name,
+				".spec.replicas can not be nil"))
+		}
+
+		if *m.Spec.Replicas <= 0 {
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mssql_webhook.go:145]: use "k8s.io/utils/ptr" ptr.Deref() and use it everywhere where you are dereferncing a pointer

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+				m.Name,
+				"number of replicas can not be less be 0 or less"))
+		}
+
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mssql_webhook.go:150]: Is it possible for AG to have a single replica?

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+		err := mssqlValidateVersion(m)
+		if err != nil {
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("version"),
+				m.Name,
+				err.Error()))
+		}
+	}
+
+	if m.IsStandalone() {
+		if *m.Spec.Replicas != 1 {
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mssql_webhook.go:127]: Give a nil check and then use "k8s.io/utils/ptr" ptr.Deref

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *MsSQL) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	mssqllog.Info("validate update", "name", m.Name)
+
+	oldMsSQL := old.(*MsSQL)
+	allErr := m.ValidateCreateOrUpdate()
+
+	if m.Spec.Topology == nil && *oldMsSQL.Spec.Replicas == 1 && *m.Spec.Replicas > 1 {
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mssql_webhook.go:83]: Use ptr.Deref

File: apis/kubedb/v1alpha2/mssql_helpers.go [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+	if err != nil {
+		klog.Errorf("can't get the MsSQL version object %s for %s \n", m.Spec.Version, err.Error())
+		return
+	}
+
+	m.setDefaultContainerSecurityContext(&mssqlVersion, m.Spec.PodTemplate)
+
+	m.SetHealthCheckerDefaults()
+
+	m.setDefaultContainerResourceLimits(m.Spec.PodTemplate)
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mssql_helpers.go:251]: set default affinity?

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_webhook.go (hunk 3 above)
Comment (Neaj-Morshad-101) [reply apis/kubedb/v1alpha2/mssql_webhook.go:150]: Yes

File: apis/catalog/v1alpha1/mssql_version_types.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	// PSP names
+	// +optional
+	PodSecurityPolicies MsSQLVersionPodSecurityPolicy `json:"podSecurityPolicies"`
+	// update constraints
+	UpdateConstraints UpdateConstraints `json:"updateConstraints,omitempty"`
+}
+
+// MsSQLVersionDatabase is the MsSQL Database image
+type MsSQLVersionDatabase struct {
+	Image string `json:"image"`
```
[the remark this answers is not in the document]
Comment (Neaj-Morshad-101) [reply apis/catalog/v1alpha1/mssql_version_types.go:78]: We don't build images for SQL server. Why do we need this?

File: apis/catalog/v1alpha1/mssql_version_types.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	// +optional
+	Stash appcat.StashAddonSpec `json:"stash,omitempty"`
+	// SecurityContext is for the additional config for the DB container
+	// +optional
+	SecurityContext MsSQLSecurityContext `json:"securityContext"`
+	// PSP names
+	// +optional
+	PodSecurityPolicies MsSQLVersionPodSecurityPolicy `json:"podSecurityPolicies"`
+	// update constraints
+	UpdateConstraints UpdateConstraints `json:"updateConstraints,omitempty"`
```
[the remark this answers is not in the document]
Comment (Neaj-Morshad-101) [reply apis/catalog/v1alpha1/mssql_version_types.go:73]: Why do we need this?

File: apis/kubedb/v1alpha2/mssql_webhook.go (hunk 4 above)
Comment (Neaj-Morshad-101) [reply apis/kubedb/v1alpha2/mssql_webhook.go:127]: If we use ptr.Deref(), why do we need the nil check? 
ptr.Deref(m.Spec.Replicas, 0) != 1 check is fine I guess.

File: apis/kubedb/v1alpha2/mysql_helpers.go [hunk 9]
```diff
@@ -372,6 +372,10 @@ func (m *MySQLSpec) GetPersistentSecrets() []string {
 	if m.AuthSecret != nil {
 		secrets = append(secrets, m.AuthSecret.Name)
 	}
+
+	secrets = append(secrets, MsSQLEndpointCertsSecretName)
+	secrets = append(secrets, MsSQLDbmLoginSecretName)
+	secrets = append(secrets, MsSQLMasterKeySecretName)
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/mysql_helpers.go:378]: Why is it here? Shouldn't it be in MSSQL Spec?

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mysql_helpers.go (hunk 9 above)
Comment (Neaj-Morshad-101) [reply apis/kubedb/v1alpha2/mysql_helpers.go:378]: Ooh!

File: apis/catalog/v1alpha1/mssql_version_types.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	InitContainer MSSQLInitContainer `json:"initContainer"`
+	// Deprecated versions usable but regarded as obsolete and best avoided, typically due to having been superseded.
+	// +optional
+	Deprecated bool `json:"deprecated,omitempty"`
+	// Stash defines backup and restore task definitions.
+	// +optional
+	Stash appcat.StashAddonSpec `json:"stash,omitempty"`
+	// SecurityContext is for the additional config for the DB container
+	// +optional
+	SecurityContext MSSQLSecurityContext `json:"securityContext"`
```
Comment (ArnobKumarSaha) [inline apis/catalog/v1alpha1/mssql_version_types.go:68]: Remove MSSQLSecurityContext, use SecurityContext

File: apis/catalog/v1alpha1/mssql_version_types.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	// PSP names
+	// +optional
+	PodSecurityPolicies MSSQLVersionPodSecurityPolicy `json:"podSecurityPolicies"`
```
Comment (ArnobKumarSaha) [inline apis/catalog/v1alpha1/mssql_version_types.go:71]: Remove PodSecurityPolicy

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_types.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+const (
+	MSSQLModeAvailabilityGroup MSSQLMode = "AvailabilityGroup"
+	MSSQLModeRemoteReplica     MSSQLMode = "RemoteReplica"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MSSQL defines a MSSQL database.
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_types.go:44]: Remove unnecessary comments

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *MSSQL) ValidateCreate() (admission.Warnings, error) {
+	mssqllog.Info("validate create", "name", m.Name)
+
+	allErr := m.ValidateCreateOrUpdate()
+	if len(allErr) == 0 {
+		return nil, nil
+	}
+	return nil, apierrors.NewInvalid(schema.GroupKind{Group: "kubedb.com", Kind: "MSSQL"}, m.Name, allErr)
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:74]: Use the constants already declared.

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+	}
+	return nil, nil
+}
+
+func (m *MSSQL) ValidateCreateOrUpdate() field.ErrorList {
+	var allErr field.ErrorList
+
+	if m.Spec.Version == "" {
+		allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("version"),
+			m.Name,
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:116]: Version is a required field.  This if statement will never be true

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+func mssqlValidateVersion(m *MSSQL) error {
+	var mssqlVersion catalog.MSSQLVersion
+	err := DefaultClient.Get(context.TODO(), types.NamespacedName{
+		Name: m.Spec.Version,
+	}, &mssqlVersion)
+	if err != nil {
+		return errors.New("version not supported")
+	}
+
+	return nil
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:225]: Just ```
return DefaultClient.Get(context.TODO(), types.NamespacedName{
		Name: m.Spec.Version,
	}, &mssqlVersion)
```

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+	mssqllog.Info("validate update", "name", m.Name)
+
+	oldMSSQL := old.(*MSSQL)
+	allErr := m.ValidateCreateOrUpdate()
+
+	if m.Spec.Topology == nil && ptr.Deref(oldMSSQL.Spec.Replicas, 0) == 1 && ptr.Deref(m.Spec.Replicas, 0) > 1 {
+		allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+			m.Name,
+			"Cannot scale up from 1 to more than 1 in standalone mode"))
+	}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:88]: Remove. Already similiar check inside m.ValidateCreateOrUpdate()

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+				m.Name,
+				"number of replicas for standalone must be one "))
+		}
+	} else {
+		if m.Spec.Topology.Mode == nil {
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("topology").Child("mode"),
+				m.Name,
+				".spec.topology.mode can't be empty in cluster mode"))
+		}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:138]: How about defaulting the mode ?

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+
+		if m.Spec.Replicas == nil {
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+				m.Name,
+				".spec.replicas can not be nil"))
+		}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:144]: Remove. The immediate next check will cover this

File: apis/kubedb/v1alpha2/mssql_webhook.go [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+
+func mssqlValidateVolumesMountPaths(podTemplate *ofst.PodTemplateSpec) error {
+	if podTemplate == nil {
+		return nil
+	}
+
+	if podTemplate.Spec.Containers == nil {
+		return nil
+	}
+
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_webhook.go:255]: Logical error. Assume a case where, User set invalid value in `podTemplate.spec.initContainers[].volumeMounts`. & Didn't set anything in `podTemplate.spec.containers`.

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_webhook.go (hunk 17 above)
Comment (Neaj-Morshad-101) [reply apis/kubedb/v1alpha2/mssql_webhook.go:125]: The default mode is Standalone now. (if spce.topology == nil)

File: apis/kubedb/v1alpha2/mssql_types.go [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+	// AvailabilityGroup info for MSSQL
+	// +optional
+	AvailabilityGroup *MSSQLAvailabilityGroupSpec `json:"availabilityGroup,omitempty"`
+}
+
+// MSSQLAvailabilityGroupSpec defines the availability group spec for MSSQL
+type MSSQLAvailabilityGroupSpec struct {
+	// AvailabilityDatabases is an array of databases to be included in the availability group
+	// +optional
+	AvailabilityDatabases []string `json:"databases"`
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/mssql_types.go:147]: Why the filed & json tag is different ?

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/mssql_types.go (hunk 20 above)
Comment (Neaj-Morshad-101) [reply apis/kubedb/v1alpha2/mssql_types.go:147]: Renamed to Databases

**Review by ArnobKumarSaha — APPROVED** [review]

**Pull request merged**